### ISSUE TITLE
Image file name standards: Why?

### ISSUE THREAD

**Issue by sbkates** [tmpworldwide/tmpworldwide.github.io#12]
I think the name of an image is less important than the name of other things.
1. The user never sees the file name in normal circumstances, even with screenreaders, as long as there is a suitable attribute such as title that the screenreader reads.
2. Developers (at least I) copy and paste the name or URL of specific images usually instead of writing from memory. This includes other developers who will be working on your code later.
3. The current suggestion assumes that an image only exists on one page. 
4. Different developers have different personal habits for this.

Personally, I like using descriptive words, such as 'woman-and-lecture-podium.jpg' for my images since that really helps me keep track of the file names while I am writing them out. That is a much better way for my mind to organize file names than something generic like '.homepage-1' which I find harder to  keep track of.

**Comment by Brockenstein** [comment]
@sylviakates If its an image that hits multiple pages wouldn't you just not state in the file name that part? and then you would just use the descriptor? Just like how you might not have multiples of the same image so you wouldn't put a number?

`page-descriptor-#.extension`

for the second image in a banner slider
`home-banner-2.jpg`

for the image in a banner of a search page with no slider
`search-banner.jpg`

for an image that hits multiple pages and there is only one of them
`social_sprite.png`

Maybe if you broke the emails out of the sprite
`social_icon-1.png`
`social_icon-2.png`
`social_icon-3.png`

**Comment by sbkates** [comment]
How is any of this easier for the developer to use, or necessary to make it easy for other developers, compared to the specific descriptions that I use? 

The developer collaborating with you is just copy and pasting the names of the files. 

It would make it harder for me to initially develop the page without being able to, at a glance, recognize that that filename is referring to the image that looks like that (woman-at-podium.jpg ). 

As a human, I don't see these filenames as homepage-1. I see it as woman-at-podium and social-icons-large. It sounds like we're trying to make these names more computer-like, like a batch process output, and I see no entity that is aided by it. Not the user, not the initial developer (in fact it confuses and slows me), not page load/ data, not the later collaborating developers.

**Comment by Brockenstein** [comment]
You can see what the image is when you click on it. When I am doing development I like images to be consistent so that I don't have to hunt down what the next file name is.

I think the overall goal of any naming convention is to make it easier for future updates. So When Bob goes to update Charlies assets he knows exactly what to look for, If everyone named their banners differently each time then it is harder to figure out what banner they need to update. I also find that woman-at-podium is confusing when there are 3 images of a woman giving speech but its different women on different pages.

Most importantly it keeps all the images consistent within one site. So that way, for my banners I don't name one woman-at-podium and the next job-banner and the next banner-245pxby4000px.

**Comment by sbkates** [comment]
"You can see what the image is when you click on it. "

Exactly. So who needs a naming standard anyway? It's not like the name of a HTML, CSS or JS file, whose content isn't clear except by putting work into reading it.

The only time I'm not seeing what an image looks like, but referring to its name, is in code. In that situation, I like being able to picture in my head which image it is that I want. "Oh the PSD shows a woman  at a podium on the home page, and a group of three guys in a meeting on the second page. Then that file named three-guys-meeting must be the file I want to put here."  

And I keep these names consistent. I don't use other ways of naming images when I'm naming them this way. 

I imagine confusion when working on a site like this. 

"Homepage-Banner-3? Wait, is that the one with the tractor, or the one with the guy in a suit?  I guess I have to keep these images open in explorer and constantly refer back to its preview function."

**Comment by howellbox** [comment]
I agree with Sylvia on this issue. I would much rather deal with woman-at-podium.jpg than home-banner-1.jpg; more importantly, having to shoehorn the page and placement metadata (which could change, or be confusing with multiple placements and/or pages) into the filename would slow down and frustrate my development process. As for updating existing code, I don't recall ever being frustrated or hampered by how someone named an image.

I appreciate the desire to keep things organized, and I think all of us would agree on its merit. Perhaps there are other options to explore? This standard sounds more like opinion than solid evidence that it would streamline development.

**Comment by designwebsean** [comment]
I agree with Sylvia here on using descriptive file names. SEO-wise, it is better to use descriptive file names. Afaik, Google would much rather have woman-at-podium.jpg over home-banner-1.jpg

**Comment by mhartz** [comment]
I actually agree with Sylvia even though it is in the standards this way.

I think the only rule should be that names must be separated by hyphens and underscores for new words.

What if an image is reused in the site? What page name does it get?

What if the image is moved from the banner to somewhere else? Do I rename it?

I actually had this marked in the checklist because of this issue. I think it is a good topic to debate.

**Comment by mhartz** [comment]
However to keep a certain naming convention, but not restricted to, for certain things should be assumed. Such as, home-banner-1.jpg and home-banner-2.jpg, or home-banner-speaker.jpg and home-banner-whiteboard.jpg.

**Comment by scampbell44** [comment]
<i>What if an image is reused in the site? What page name does it get? What if the image is moved from the banner to somewhere else? Do I rename it? </i>

<i>I am thinking a more stylized approach is better. Eg. Use dashes to separate words and underscore to separate word groups.</i>

Excellent points! I concur with using descriptive image file names instead of placement based names and specifying standards in the stylistic elements such as the use of dashes and underscores.

**Comment by scampbell44** [comment]
oops.. clicked the wrong button and didn't mean to close this. How do i re-open?

**Comment by sbkates** [comment]
How do I upvote because I want to upvote Max Hartz's posts.

**Comment by michaelspellacy** [comment]
@designweb1 makes a good point about SEO. I don't particularity care for most SEO tactics, but in being searchable on the web, giving images descriptive names is a good thing. Good post, @sylviakates.

Also good from an accessibility standpoint. While screen readers ignore null alts, some readers do allow you to configure them to read out the name of the graphic, so it is a win there as well. ;-)

Crazy idea: For us programmers, how about using more function or layout specific class or data attributes? 

`<img src="board-room-meeting.jpg" alt="" data-page-reference="home">`

Just a thought.

**Comment by mhartz** [comment]
So I think it is pretty unanimous that we should be using descriptive image file names. 

What does everyone think about  using hyphens for words and underscores for word groups, or vice versa? Or should we just use one or the other?

Hyphens for words, underscores for word groups:
`<image src="board-room-meeting_small.jpg" alt="Board Room Meeting">`

Underscores for words, hyphens for word groups:
`<image src="board_room_meeting-small.jpg" alt="Board Room Meeting">`

Just hyphens:
`<image src="board-room-meeting-small.jpg" alt="Board Room Meeting">`

Just Underscores:
`<image src="board_room_meeting_small.jpg" alt="Board Room Meeting">`

I kind of like the underscore for words and hyphens for word groups because of one reason: double click or Ctrl+shift selects the word groups even though I don't think it looks as good. Anyone else have an opinion?

Also, should we be using data attributes for anything other than as javascript references? Or are they ok for CSS hooks as well?

**Comment by michaelspellacy** [comment]
> > Also, should we be using data attributes for anything other than as javascript references? Or are they ok for CSS hooks as well?

Probably better not to have, now that I think about it a little more.

I like option 3, just hyphens. Going back to being searchable, it is consistent with how we should be naming pages: web-standards.html. I like that consistency across the board. One file naming style to rule them all. :-)

**Comment by michaelspellacy** [comment]
They could be used for "hooks", but I don't think it is their intended purpose. Ever read this? http://danwebb.net/2010/1/27/put-that-data-attribute-away-son-you-might-hurt-someone

**Comment by mhartz** [comment]
I also have this small fear that you could accidentally name your data attribute with one that has a hook in a 3rd party library by accident. Who knows how many "data-content"s there are out there.

But I am fine with hyphens.  Curious to see what other people think.

**Comment by scampbell44** [comment]
Agreed on both ... 

Data attributes not being intended as hooks has the potential for more issues than not. I say reserve these for js purposes.

Personally, i prefer just hyphens. I think it's cleaner and easier to read.

**Comment by mhartz** [comment]
I'm starting to wonder if we should even set such a strict rule as hyphens only. Did a little vote in the office and it was a mix between hyphens only vs. hyphens for words and underscores/hyphens for words and sections. I'm also thinking that the file naming conventions should apply to all files instead of having one for some files and another just for images.

**Comment by stephendelancey** [comment]
I am with Max, it should also be for all files.

**Comment by sbkates** [comment]
A note, I think that image names are different from other file names since the developer can easily understand an image's contents at just a glance, but JS, CSS files' content have to be read in detail, in a time consuming way, to understand them. For this reason, the non-image files need a more information-rich title and therefore, stricter rules on their names.

**Comment by mhartz** [comment]
I am in the same boat personally, but I am trying to stir the pot to get a debate going.

Here are my personal thoughts on the matter:

I like hyphens only for images and hyphen/underscores for other files names. Image names basically need to be descriptive of what the image contains, however other files have different uses.

Think about the scenario where we are hosting a bunch of PDF newsletters. They have the same name, but each one has a different date. For example, "company-newsletter-2015-05-04.pdf" would be better represented as, "company-newsletter_2015-05-04.pdf" or even as, "company_newsletter-2015_05_04.pdf". There are other scenarios too such as file versioning as well.

Also, I personally like to force certain files to the top of directories by prepending them with an underscore, eg. "_reset.scss".

Let me know what you think, and everyone would be cool with just hyphens on images, we can open a new issue for all other files and close this one with a new standard for just images.

**Comment by michaelspellacy** [comment]
I still say hyphens across the board for images, page names or any other asset used to make up construction of site/page. I hate to cite SEO again but...https://www.youtube.com/watch?t=14&v=AQcSFsQyct8

Fine with using underscore outside of any non-index-able files, like Sass.

**Comment by mhartz** [comment]
Watched the video and it makes sense. The basic idea is that hyphens are used by Google to index each word, while underscores either lumps them together as one word or includes the underscore itself the search term.

So "word1-word2" would return results when "word1", "word2", "word1-word2", or "word1 word2" are entered in the search input. However, "word1_word2" would only return results on "word1_word2" and "word1word2", which would very rarely be the case.

I'm in on hyphens only.

If anyone else has any input or other articles to dispute this, please chime in.

**Comment by michaelspellacy** [comment]
:+1:

**Comment by mhartz** [comment]
Ran it through our office and everyone is on board for hyphens only. If no one else disagrees I will add this at the end of the day.

**Comment by scampbell44** [comment]
<blockquote>So "word1-word2" would return results when "word1", "word2", "word1-word2", or "word1 word2" are entered in the search input. However, "word1_word2" would only return results on "word1_word2" and "word1word2", which would very rarely be the case.</blockquote>

Excellent examples Max. This makes the argument for hyphens self-evident.

**Comment by mhartz** [comment]
Here is the new proposed change to the documentation, all other naming convention documentation will be removed:
1.5 File Naming Conventions
- Only use lowercase alphanumeric characters and hyphens.
  - Non-indexable files (such as SASS) may also contain underscores.
- Names should be descriptive of the file's contents or image.

```
<!-- Not recommended -->
Banner Home Background.jpg
logo_1.png
doc-How_To_Apply.pdf
coffee.shop.jpg

<!-- Recommended -->
home-banner-background.jpg
logo-1.png
doc-how-to-apply.pdf
coffee-shop.jpg
```

**Comment by mhartz** [comment]
Added to the standards and issue closed in commit 0f230224cd8c3af885b2cce9dc58fd796c5b85c1
